openedx-unsupported/edx-analytics-dashboard
# Add page that shows problem history for a single user and course

This PR adds a new page to the Insights dashboard that shows, for a specific user and course, the list of problems that the user has attempted, along with the total number of attempts and the final (i.e., most recent) score.

The new page can be accessed by choosing the "Problem Data" entry from the dropdown menu that is available in the navigation bar while viewing a user's profile. (The dropdown menu also provides a "User Profile" entry that allows to switch back to a user's profile after viewing problem information on the new page.)

While viewing problem data the dashboard is in "user mode": The course mode selector can be used to enable course mode and jump back to the home page of the parent course; the user mode selector points to the list of all users currently enrolled in the parent course. (See https://github.com/edx/edx-analytics-dashboard/pull/338 for more information about mode selectors, user mode, and course mode.)
### Screenshots

New page listing problems user has attempted:

![user-problem-history](https://cloud.githubusercontent.com/assets/961441/9408105/d57cb172-480f-11e5-900d-78298f748aa3.png)

The dashboard uses the Course Structure API to obtain display names for problems when generating this page. If the `course_api_enabled` switch is off, it falls back on problem IDs:

![user-problem-history-course-api-disabled](https://cloud.githubusercontent.com/assets/961441/9519915/4600a66c-4cc3-11e5-9b6c-02d921068e33.png)

Navigation lens on new page:

![navigation-lens-on-problem-history-page](https://cloud.githubusercontent.com/assets/961441/9408185/833ddd0e-4810-11e5-840d-9ce448f98b75.png)

Navigation lens on profile page:

![navigation-lens-on-profile-page](https://cloud.githubusercontent.com/assets/961441/9408187/8e42ff36-4810-11e5-8e11-60bfc641146f.png)
### Dependencies
- https://github.com/edx/edx-analytics-pipeline/pull/147
- https://github.com/edx/edx-analytics-data-api/pull/81
- https://github.com/edx/edx-analytics-data-api-client/pull/22
- https://github.com/edx/edx-analytics-dashboard/pull/338

Note that https://github.com/edx/edx-analytics-dashboard/pull/338 has data-api and data-api-client dependencies of its own. For the purpose of testing this PR you will therefore need to work with branches that contain relevant changes for both https://github.com/edx/edx-analytics-dashboard/pull/338 and this PR. You won't need to create these branches yourself; they have already been set up for you (cf. test instructions below).
### Testing
1. Test https://github.com/edx/edx-analytics-pipeline/pull/147 and confirm it's working.
2. Prepare the **data-api** repository by switching to [this branch](https://github.com/open-craft/edx-analytics-data-api/tree/user-problem-data) (`user-problem-data`).
3. Prepare the **data-api-client** repository by switching to [this branch](https://github.com/open-craft/edx-analytics-data-api-client/tree/user-problem-data) (`user-problem-data`). 
4. Install the updated data API client into the dashboard's virtualenv. Use these commands (as `analytics` user):
   
   ```
   cd ~/apps/data-api-client/
   deactivate
   source ~/venvs/dashboard/bin/activate
   pip install -e .
   ```
5. Start the data API server (as `analytics` user):
   
   ```
   cd ~/apps/data-api/ && ./manage.py runserver 0.0.0.0:9001
   ```
6. Start the dashboard (as `analytics` user):
   
   ```
   cd ~/apps/dashboard/ && ./manage.py runserver 0.0.0.0:9999
   ```
7. Access the Insights page for the demo course at http://192.168.33.11:9999/courses/edX/DemoX/Demo_Course/.
8. Follow these steps to try out the new functionality:
   - Click the user mode selector to switch to user mode and access the list of enrolled students.
   - Click the username of a student to access their profile.
   - Expand the dropdown menu in the navigation bar and click "Problem Data" to access the new page.
### Partner Information

Not an edX partner - 3rd party-hosted open edX instance
